sdk/ai/Azure.AI.Inference/src/Telemetry/OpenTelemetryScope.cs

```diff
@@ -0,0 +1,255 @@
+                {
+                    TagList requestTags = new() {
+                    { GenAiSystemKey, GenAiSystemValue},
+                    { GenAiEventContent, GetContent(message)}
+                };
+                    _activity?.AddEvent(
+                        new ActivityEvent(
+                            $"gen_ai.{message.Role}.message",
+                            default,
+                            new ActivityTagsCollection(requestTags)
+                        )
+                    );
+                }
+            }
+            _duration = Stopwatch.StartNew();
+        }
+
+        public void RecordResponse(ChatCompletions response)
+        {
+            RecordResponseInternal(new SingleRecordedResponse(response, _traceContent));
+        }
+
+        public void UpdateStreamResponse<T>(T item)
+        {
+            if (!_enableTelemetry || item is not StreamingChatCompletionsUpdate)
+                return;
+            StreamingChatCompletionsUpdate castedItem = item as StreamingChatCompletionsUpdate;
+            m_recordedStreamingResponse.Update(castedItem);
+        }
+
+        public void RecordStreamingResponse()
+        {
+            RecordResponseInternal(m_recordedStreamingResponse);
+        }
+
+        /// <summary>
+        /// Log the error.
+        /// </summary>
+        /// <param name="e">Exception thrown by completion call.</param>
+        public void RecordError(Exception e)
+        {
+            if (!_enableTelemetry)
+                return;
+            if (!string.IsNullOrEmpty(_caller))
+                m_source.Write(_caller + ".Exception", e);
+            var errorType = e?.GetType()?.FullName;
+            var tags = _commonTags;
+            tags.Add(ErrorTypeKey, errorType);
+            s_duration.Record(_duration.Elapsed.TotalSeconds, tags);
+            if ( e is Azure.RequestFailedException requestFailed)
+            {
+                errorType = requestFailed.Status.ToString();
+            }
+            _activity?.SetTag(ErrorTypeKey, errorType);
+            _activity?.SetStatus(ActivityStatusCode.Error, e?.Message ?? errorType);
+        }
+
+        /// <summary>
+        /// Record the events and metrics associated with the response.
+        /// </summary>
+        /// <param name="recordedResponse"></param>
+        private void RecordResponseInternal(AbstractRecordedResponse recordedResponse)
+        {
+            if (!_enableTelemetry || recordedResponse.IsEmpty)
+                return;
+            TagList tags = _commonTags;
+            // Find index of model tag.
+            object objOldModel = _activity?.GetTagItem(GenAiResponseModelKey);
+            if (objOldModel != null && !objOldModel.ToString().Equals(recordedResponse.Model))
+            {
+                tags.Remove(new KeyValuePair<string, object> (GenAiResponseModelKey, objOldModel));
+            }
+            tags.Add(GenAiResponseModelKey, recordedResponse.Model);
```

Comment:
this doesn't look right. 

Response model returned in all chunks should be the same. Please write the code in the way that it's added once and is not updated/removed later.

Reply:
RecordResponseInternal is being called only once, when the all the data was already collected in case of streaming scenario (in `finally` block) and when the request is completed in basic scenario. The is block is only needed because sometimes model name is not the same in request and in response.